core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,705 @@
+            this.partitionedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(partitionedExchangeSourceHandles, "partitionedExchangeSourceHandles is null"));
+            this.replicatedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(replicatedExchangeSourceHandles, "replicatedExchangeSourceHandles is null"));
+            this.splitBatchSize = splitBatchSize;
+            this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
+            this.bucketToPartitionMap = requireNonNull(bucketToPartitionMap, "bucketToPartitionMap is null");
+            checkArgument(bucketToPartitionMap.isPresent() || partitionedExchangeSourceHandles.isEmpty() || splitSources.isEmpty(), "bucket to partition map is expected to be set when partitioned inputs are present");
+            this.bucketNodeMap = requireNonNull(bucketNodeMap, "bucketNodeMap is null");
+            checkArgument(splitSources.isEmpty() || bucketNodeMap.isPresent(), "splitToBucketFunction must be set when split sources are present");
+            this.catalogRequirement = requireNonNull(catalogRequirement, "catalogRequirement is null");
+        }
+
+        @Override
+        public List<TaskDescriptor> getMoreTasks()
+        {
+            if (finished || closed) {
+                return ImmutableList.of();
+            }
+
+            Map<Integer, Multimap<PlanNodeId, Split>> partitionToSplitsMap = new HashMap<>();
+            Map<Integer, HostAddress> partitionToNodeMap = new HashMap<>();
+            if (!splitSources.isEmpty()) {
+                for (Map.Entry<PlanNodeId, SplitSource> entry : splitSources.entrySet()) {
+                    PlanNodeId scanNodeId = entry.getKey();
+                    SplitSource splitSource = entry.getValue();
+                    BucketNodeMap bucketNodeMap = this.bucketNodeMap
+                            .orElseThrow(() -> new VerifyException("bucket to node map is expected to be present"));
+                    while (!splitSource.isFinished()) {
+                        ListenableFuture<SplitBatch> splitBatchFuture = splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize);
+
+                        long start = System.nanoTime();
+                        addSuccessCallback(splitBatchFuture, () -> getSplitTimeRecorder.accept(start));
+
+                        SplitBatch splitBatch = getFutureValue(splitBatchFuture);
+
+                        for (Split split : splitBatch.getSplits()) {
+                            int bucket = bucketNodeMap.getBucket(split);
+                            int partition = bucketToPartitionMap.map(map -> map[bucket]).orElse(bucket);
+
+                            if (!bucketNodeMap.isDynamic()) {
+                                partitionToNodeMap.put(partition, bucketNodeMap.getAssignedNode(split).get().getHostAndPort());
```

Comment:
It is not obvious we are not overriding entry in map here. Add a check that we are not (or that we are not changing the value)